src/main/java/org/gaul/s3proxy/CrossOriginResourceSharing.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2014-2018 Andrew Gaul <andrew@gaul.org>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gaul.s3proxy;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.Strings;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class CrossOriginResourceSharing {
+    private static final String VALUE_SEPARATOR = " ";
+    private static final String HEADER_VALUE_SEPARATOR = ", ";
+    private static final String ALLOW_ANY_HEADER = "*";
+
+    private static final Logger logger = LoggerFactory.getLogger(
+            CrossOriginResourceSharing.class);
+
+    private String allowedMethodsRaw;
+    private String allowedHeadersRaw;
+    private List<Pattern> allowedOrigins;
+    private List<String> allowedMethods;
+    private List<String> allowedHeaders;
+
+    protected CrossOriginResourceSharing() {
+        // CORS Allow all
+        this(".+", "GET PUT POST", ALLOW_ANY_HEADER);
+    }
+
+    protected CrossOriginResourceSharing(String allowedOrigins,
+            String allowedMethods, String allowedHeaders) {
+        this.allowedOrigins = new ArrayList<Pattern>();
+        this.allowedMethods = new ArrayList<String>();
+        this.allowedHeaders = new ArrayList<String>();
+
+        if (!Strings.isNullOrEmpty(allowedOrigins)) {
+            for (String origin: allowedOrigins.split(VALUE_SEPARATOR)) {
+                this.allowedOrigins.add(
+                        Pattern.compile(origin, Pattern.CASE_INSENSITIVE));
+            }
+        }
+        if (!Strings.isNullOrEmpty(allowedMethods)) {
+            this.allowedMethods.addAll(Arrays.asList(allowedMethods.split(
+                    VALUE_SEPARATOR)));
+        }
+        if (!Strings.isNullOrEmpty(allowedMethods)) {
+            this.allowedHeaders.addAll(Arrays.asList(allowedHeaders.split(
```

Comment:
Simplify to `this.allowedHeaders = ImmutableSet.copyOf(allowedHeaders.split(VALUE_SEPARATOR))`.